src/databricks/labs/ucx/source_code/files.py

```diff
@@ -44,11 +50,93 @@ def __repr__(self):
+        self._path = path
+        self._file_loader = file_loader
+        self._dir_loader = dir_loader
+
+    @property
+    def path(self) -> Path:
+        return self._path
+
+    def build_dependency_graph(self, parent: DependencyGraph) -> list[DependencyProblem]:
+        return list(self._build_dependency_graph(parent))
+
+    def _build_dependency_graph(self, parent: DependencyGraph) -> Iterable[DependencyProblem]:
+        for child_path in self._path.iterdir():
+            if child_path.is_dir():
+                dependency = Dependency(self._dir_loader, child_path)
+                yield from parent.register_dependency(dependency).problems
+                continue
+            dependency = Dependency(self._file_loader, child_path)
+            yield from parent.register_dependency(dependency).problems
+
+    def __repr__(self):
+        return f"<LocalDirectory {self._path}>"
+
+
+class LocalFilesLinter:
+
+    def __init__(
+        self,
+        languages: Languages,
+        file_loader: FileLoader,
+        dir_loader: DirectoryLoader,
+        path_lookup: PathLookup,
+        dependency_resolver: DependencyResolver,
+    ) -> None:
+        self._languages = languages
+        self._file_loader = file_loader
+        self._dir_loader = dir_loader
+        self._path_lookup = path_lookup
+        self._dependency_resolver = dependency_resolver
+        self._extensions = {".py": Language.PYTHON, ".sql": Language.SQL}
+
+    def lint(self, path: Path) -> list[LocatedAdvice]:
+        return list(self._lint(path))
+
+    def _lint(self, path: Path) -> Iterable[LocatedAdvice]:
+        loader = self._dir_loader if path.is_dir() else self._file_loader
+        dependency = Dependency(loader, path)
+        graph = DependencyGraph(dependency, None, self._dependency_resolver, self._path_lookup)
+        container = dependency.load(self._path_lookup)
+        assert container is not None  # because we just created it
+        dependency_problems = container.build_dependency_graph(graph)
+        for dependency_problem in dependency_problems:
+            advice = self._located_advice_from_dependency_problem(dependency_problem)
+            yield advice
+        for child_path in graph.all_paths:
+            yield from self._lint_one(child_path)
+
+    @staticmethod
+    def _located_advice_from_dependency_problem(problem: DependencyProblem):
+        return LocatedAdvice(
+            path=problem.source_path.absolute() if problem.source_path.name != MISSING_SOURCE_PATH else Path('UNKNOWN'),
+            code=problem.code,
+            message=problem.message,
+            start_line=problem.start_line,
+            start_col=problem.start_col,
+            end_line=problem.end_line,
+            end_col=problem.end_col,
+        )
+
+    def _lint_one(self, path: Path) -> Iterable[LocatedAdvice]:
+        if path.is_dir():
+            return []
+        linter = FileLinter(self._languages, path)
```

Comment:
```suggestion
        session_state = CurrentSessionState()
        languages = Languages(self._migration_index, session_state)
        linter = FileLinter(languages, path)
```

Reply:
now using a LanguagesFactory to avoid spreading those details